### PR TITLE
refactor: disallow parametrized items in auto menu

### LINKED ISSUE #2378: createMenuItems includes parameterized routes
### Describe the bug

The menu items created by `createMenuItems` includes paths with parameters, leading to links like `users/{id}/edit` in the menu
image_720

### Expected-behavior

Only paths without parameters are shown

### Reproduction

Create the following structure in a Hilla starter `users/{id}/edit.tsx`

### System Info

Hilla 24.4.0.alpha25

### PR DESCRIPTION
**Pull request by Lodin**
Fixes #2378.

### CODE REVIEW

File: packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java [hunk 1]
```diff
@@ -142,18 +140,21 @@ protected Map<String, AvailableViewInfo> collectClientViews(
...
-                .filter(clientViewConfigEntry -> routeUtil.isRouteAllowed(
-                        isUserInRole, isUserAuthenticated,
-                        clientViewConfigEntry.getValue()))
-                .forEach(clientViewConfigEntry -> {
+                .filter(clientViewConfigEntry -> {
+                    var config = clientViewConfigEntry.getValue();
+                    return routeUtil.isRouteAllowed(isUserInRole,
+                            isUserAuthenticated, config)
+                            && (config.getRouteParameters() == null
+                                    || config.getRouteParameters().isEmpty());
```
Comment (taefi) [inline packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java:148]: I would do this in the following order/style:
1. map clientViewConfigEntry -> clientViewConfigEntry.getValue()
2. filter those satisfying `(config.getRouteParameters() == null || config.getRouteParameters().isEmpty())`
3. filter those satisfying `routeUtil.isRouteAllowed(isUserInRole, isUserAuthenticated, config)`

Comment (taefi) [reply packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java:148]: I would also replace the `.forEach` part with 
```java
.collect(Collectors.toMap(AvailableViewInfo::route, Function.identity()))
```

File: packages/ts/file-router/src/shared/internal.d.ts [hunk 2]
```diff
@@ -11,13 +10,8 @@ export type ServerViewConfig = Readonly<{
...
   ViewConfig;
 
-export type ServerViewMapItem = Readonly<{
-  params?: Readonly<Record<string, RouteParamType>>;
-}> &
-  ViewConfig;
-
 export type VaadinObject = Readonly<{
-  views: Readonly<Record<string, ServerViewMapItem>>;
+  views: Readonly<Record<string, ViewConfig>>;
```
Comment (taefi) [inline packages/ts/file-router/src/shared/internal.d.ts:14]: I still wonder whether the views with optional parameters should be available as menus.

**Review by taefi — CHANGES_REQUESTED** [review]

File: packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java (hunk 1 above)
Comment (taefi) [inline packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java:148]: According to [this](https://github.com/vaadin/hilla/issues/2378#issuecomment-2096002474) comment, the `(config.getRouteParameters() == null || config.getRouteParameters().isEmpty())` which exclude views with parameters (either optional or required) should change to only exclude the routes with required parameters.

**Review by taefi — CHANGES_REQUESTED** [review]

File: packages/ts/file-router/test/runtime/createMenuItems.spec.ts [hunk 3]
```diff
@@ -33,10 +32,7 @@ describe('@vaadin/hilla-file-router', () => {
           '/profile/account/security/password': { title: 'Password' },
           '/profile/account/security/two-factor-auth': { title: 'Two Factor Auth' },
           '/profile/friends/list': { title: 'List' },
-          '/profile/friends/:user': { title: 'User', params: { ':user': RouteParamType.Required } },
           '/test/empty': {},
```
Comment (platosha) [inline packages/ts/file-router/test/runtime/createMenuItems.spec.ts:35]:
```suggestion
          '/test/empty': {},
          '/test/:optional?': { title: 'Optional', params: { ':optional?': RouteParamType.Optional } },
          '/test/*': { title: 'Wildcard', params: { '*': RouteParamType.Wildcard } },
```

File: packages/ts/file-router/test/runtime/createMenuItems.spec.ts [hunk 4]
```diff
@@ -63,17 +59,9 @@ describe('@vaadin/hilla-file-router', () => {
           title: 'List',
           to: '/profile/friends/list',
         },
```
Comment (platosha) [inline packages/ts/file-router/test/runtime/createMenuItems.spec.ts:61]:
```suggestion
        },
        {
          title: 'Optional',
          to: '/test/',
        },
```

File: packages/ts/file-router/test/runtime/createMenuItems.spec.ts [hunk 5]
```diff
@@ -63,17 +59,9 @@ describe('@vaadin/hilla-file-router', () => {
...
-        {
-          title: 'Optional',
-          to: '/test/',
-        },
         {
           to: '/test/empty',
         },
```
Comment (platosha) [inline packages/ts/file-router/test/runtime/createMenuItems.spec.ts:64]:
```suggestion
        },
        {
          title: 'Wildcard',
          to: '/test/',
        },
```

File: packages/ts/file-router/test/runtime/createMenuItems.spec.ts [hunk 6]
```diff
@@ -92,11 +80,8 @@ describe('@vaadin/hilla-file-router', () => {
           '/profile/account/security/password': { title: 'Password', menu: { order: 20 } },
           '/profile/account/security/two-factor-auth': { title: 'Two Factor Auth', menu: { order: 20 } },
           '/b': { title: 'List' },
-          '/profile/friends/:user': { title: 'User', params: { ':user': RouteParamType.Required } },
           '/': { title: 'Root' },
           '/test/empty': { title: 'empty', menu: { order: 5 } },
```
Comment (platosha) [inline packages/ts/file-router/test/runtime/createMenuItems.spec.ts:84]:
```suggestion
          '/test/empty': { title: 'empty', menu: { order: 5 } },
          '/test/:optional?': { title: 'Optional', params: { ':optional?': RouteParamType.Optional } },
          '/a/*': { title: 'Wildcard', params: { '*': RouteParamType.Wildcard } },
```

File: packages/ts/file-router/test/runtime/createMenuItems.spec.ts [hunk 7]
```diff
@@ -130,10 +115,6 @@ describe('@vaadin/hilla-file-router', () => {
           title: 'Root',
           to: '/',
         },
```
Comment (platosha) [inline packages/ts/file-router/test/runtime/createMenuItems.spec.ts:117]:
```suggestion
        },
        {
          title: 'Wildcard',
          to: '/a/',
        },
```

File: packages/ts/file-router/test/runtime/createMenuItems.spec.ts [hunk 8]
```diff
@@ -142,10 +123,6 @@ describe('@vaadin/hilla-file-router', () => {
           title: 'List',
           to: '/b',
         },
```
Comment (platosha) [inline packages/ts/file-router/test/runtime/createMenuItems.spec.ts:125]:
```suggestion
        },
        {
          title: 'Optional',
          to: '/test/',
        },
```

**Review by platosha — CHANGES_REQUESTED** [review]
Let's keep the test fixtures for optional and wildcard parameters, as they are (kind of) supported use cases.

**Review by taefi — COMMENTED** [review]
From the code changes POV it LGTM!
From the tests POV, I agree with @platosha's comment regarding the support of views with optional params. So, I'm not going to restate those requested changes.

**Review by platosha — COMMENTED** [review]
Upon a closer look, there is still an issue that the link target path contains optional segments, such as `/:___commentId?` in `/comments/:___commentId?`, for example. Somewhere we need to omit them...

**Comment by platosha** [comment]
> Upon a closer look, there is still an issue that the link target path contains optional segments, such as `/:___commentId?` in `/comments/:___commentId?`, for example. Somewhere we need to omit them...

addressed in [ba2a093](https://github.com/vaadin/hilla/pull/2392/commits/ba2a09361cf332ad1261b6a65b8b8e4411d8b010)

**Review by platosha — APPROVED** [review]

**Comment by platosha** [comment]
> From the code changes POV it LGTM! From the tests POV, I agree with @platosha's comment regarding the support of views with optional params. So, I'm not going to restate those requested changes.

I went with the tactic to omit the segments in Java, as it is more convenient there having the parameter info. Now that `createMenuItems` has nothing to do with route parameters, it just uses the path key as is — omitting tests for the use cases is fine.

**Review by taefi — APPROVED** [review]

**Pull request merged**